keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

```diff
@@ -0,0 +1,363 @@
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: Action
+metadata:
+  name: action-sample
+spec:
+  upgrade: 
+      targetVersion: v1.16.0
+```
+
+or 
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: Action
+metadata:
+  name: action-sample
+spec:
+  renewCertificates: {}
+```
+#### Action Controller and TaskGroup
+The `Action controller` is a component of the kubeadm operator that manages the  `Action` CRD and implements all the domain-specific knowledge about the kubeadm workflows e.g. it knows that the certificate-renewal workflow requires to run `kubeadm alpha certs renew` on all the control plane nodes in the cluster.
+
+As a consequence, the  `Action controller`  generates one or more `TaskGroup` objects, each describing a `TaskTemplate` object - e.g. run `kubeadm alpha certs renew` -  targeting a set of nodes matching a given `NodeSelector` - e.g. all nodes with `node-role.kubernetes.io/master` label -.
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: TaskGroup
+metadata:
+  name: taskgroup-sample2
+  ownerReference:
+      ...
+spec:
+  nodeSelector: 
+    matchLabels:
+      node-role.kubernetes.io/master: ""
+  selector:
+    matchLabels:
+      app: a
+  template:
+    metadata:
+      labels:
+        app: a
+    alphaCertsRenew: {}
+```
+
+It is important to notice that the  `Action controller` is going to always generate `TaskGroup` in sequential order - only after the current `TaskGroup` is completed successfully a new one will be generated.
+
+> during implementation of the first proof of concept we might consider to extend `nodeSelector` semantic in order to express criteria like e.g. the first control plane node
+
+> during implementation of the first proof of concept we might consider to extend `Task` template semantic in order to allow to execute more than one command.
+
+> during implementation of the first proof of concept we might consider to implement a lock mechanisms for prevventing concurrent execution of more than one `Action` at any time.
+
+> Alternative name for the `TaskGroup` might be `TaskBatch` (batch = a quantity or consignment of goods produced at one time).
+> Instead usage of `TaskSet` or `TaskDeployment` was discarded because of its generally imply something that creates Pods, which it is not the case.
+
+
+#### TaskGroupController and Task
+The `TaskGroup controller` will be responsible for managing the  `TaskGroup` CRD and implements all the logic for generating `Task` from a `TaskGroup` according to the current topology of the Kubernetes cluster.
+
+Each generated `Task` is expected to target a specific `Node` in the cluster.
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: Task
+metadata:
+  name: task-sample-control-plane-1
+  labels:
+    app: a
+spec:
+  nodeName: control-plane-1
+  alphaCertsRenew: {}
```

Comment:
Does `alphaCertsRenew` include any fields ?

Maybe we should add the types in the design

Reply:
I don't have a design ready for all the Task, but I assume they will be similar to the command CLI.

However, at this stage, I prefer to get a high-level agreement on how the operator is goin to orchestrate workflows (with `Action`, `TaskGroup` and `Task`) before getting into the detailed API review

Reply:
sgtm